c/cert/src/rules/INT34-C/ExprShiftedbyNegativeOrGreaterPrecisionOperand.ql

```diff
@@ -0,0 +1,134 @@
+/**
+ * @id c/cert/expr-shiftedby-negative-or-greater-precision-operand
+ * @name INT34-C: Bit shift should not be done by a negative operand or an operand of greater-or-equal precision than that of another
+ * @description Shifting an expression by an operand that is negative or of precision greater or
+ *              equal to that or the another causes representational error.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/cert/id/int34-c
+ *       external/cert/obligation/rule
+ */
+
+import cpp
+import codingstandards.c.cert
+import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
+import semmle.code.cpp.ir.internal.ASTValueNumbering
+import semmle.code.cpp.controlflow.Guards
+
+/*
+ * Precision predicate based on a sample implementation from
+ * https://wiki.sei.cmu.edu/confluence/display/c/INT35-C.+Use+correct+integer+precisions
+ */
+
+/**
+ * A function whose name is suggestive that it counts the number of bits set.
+ */
+class PopCount extends Function {
+  PopCount() { this.getName().toLowerCase().matches("%popc%nt%") }
+}
+
+/**
+ * A macro which is suggestive that it is used to determine the precision of an integer.
+ */
+class PrecisionMacro extends Macro {
+  PrecisionMacro() { this.getName().toLowerCase().matches("precision") }
+}
+
+class LiteralZero extends Literal {
+  LiteralZero() { this.getValue() = "0" }
+}
+
+class BitShiftExpr extends BinaryBitwiseOperation {
+  BitShiftExpr() {
+    this instanceof LShiftExpr or
+    this instanceof RShiftExpr
+  }
+
+  override string toString() {
+    if this instanceof LShiftExpr then result = "left-shift" else result = "right-shift"
+  }
+}
+
+int getPrecision(BuiltInType type) {
```

Comment:
This can be simplified to:
```
int getPrecision(IntegralType type) {
  type.isExplicitlyUnsigned() and result = type.getSize() * 8
  or
  type.isExplicitlySigned() and result = (type.getSize() * 8) - 1
}
```

Reply:
👍